### ISSUE TITLE
Documents recursively recording

### ISSUE THREAD

**Issue by SeeDborg** [OpenBoard-org/OpenBoard#850]
Hello,
I changed the setting in the config file : 

DataDirectory=./library/documents

As a result, my documents are indeed saved here but I noticed that they recursively create folders, for example : 
\library\documents\document\OpenBoard Document 2023-12-13 00-35-24.147\library\documents\document\OpenBoard Document 2023-12-13 00-35-24.147

It's just a small example, sometimes I have thousands of folders inside folders inside folders ! And it's taking a large amount on the hard drive.

**Comment by letsfindaway** [comment]
Please use an absolute path for the DataDirectory setting.

However I agree, this should not happen and a relative path should also be possible.

**Comment by SeeDborg** [comment]
I'll use an absolute path for now but will you try to fix this soon ?
My aim was to put my entire OpenBoard folder inside of my dropbox so that I can use OpenBoard at home or at work on 2 different computers.

I became aware of the issue because I had dropbox syncs all the time, that took forever ^^

**Comment by SeeDborg** [comment]
Hi, is his problem fixed in 1.7.1 ?
It nearly destroyed my SSD so I want to be cautious before using it next time 😅

**Comment by letsfindaway** [comment]
> I'll use an absolute path for now but will you try to fix this soon ?

When using a relative path, what should be the base directory for OpenBoard?

- The user application directory, i.e. the directory where normally the `document` directory and the `OpenBoardUser.config` is located?
- The current working directory? However most users are not aware what the current working directory is when they start an application by clicking an icon.
- Anything else?

**Comment by SeeDborg** [comment]
The most logical, in my opinion, would be that the base directory is the directory where OpenBoard is installed.

The user application directory is not used by anyone and the openboard directory is in a hidden subfolder so no one know it's there !

**Comment by letsfindaway** [comment]
I think the installation directory is not the right choice.

- On Windows it is somewhere in `C:\Program Files`. This hierarchy should not contain user data files. They should be in `C:\Users\<USER>`.
- On Linux, it is either in `/opt` or `/usr`. Both directories and all subdirectories are not writable by a normal user. User data files should be somewhere below `/home/<user>`.

But let me phrase the question the other way: What is the benefit of specifying a relative path? I don't see the value.

**Comment by SeeDborg** [comment]
It's very simple : it would allow me to have a portable version of Openboard, that I can put on a USB key for the substitute teachers, as stated in [#668](url)

**Comment by letsfindaway** [comment]
Here an idea for this:

Start OpenBoard from the stick using a script. This script can get the path to the script as `%~dp0`; see https://superuser.com/questions/1652732/how-to-get-the-directory-path-where-a-batch-file-is-called-from. The script can then modify the `OpenBoardUser.config` file accordingly before actually starting OpenBoard. I think the PowerShell replace command could be used for this purpose.

Using this technique we can guarantee that the `DataDirectory` is always an absolute path and that the correct drive letter is included.

For Linux, I found a small helper program for openSUSE here: https://build.opensuse.org/package/show/home%3Afschuett/schule-openboard When it is started, it replaces the OpenBoard document directory by a symlink to an USB stick as soon as it is inserted and restores the previous configuration once it is removed.

**Comment by SeeDborg** [comment]
I'm sorry but for now, I'm not an advanced enough user to get this working 😅

I really love your project but I cant really see why you don't simply offer a portable version aside of the installable version. Nowadays, almost every software on github is doing that.

Putting the document folders inside the installation folder would be the more logical solution : no one in schools use openboard for different users on the same computer.
Or at least, adding an option to choose the document folder would be great.

**Comment by kaamui** [comment]
> I really love your project but I cant really see why you don't simply offer a portable version aside of the installable version.

@letsfindaway is just a volunteer that helps me maintain the project and is not to blame. I'm the maintainer and I don't offer a portable version because it's not a priority at all, not asked by the users I'm paid to work for, aka Geneva teachers. 

> Putting the document folders inside the installation folder would be the more logical solution : no one in schools use openboard for different users on the same computer.

To put the OpenBoard documents inside the installation directory is simply a bad practice, especially on Windows, where Program Files folder needs admin privileges to be able to write in. The good practice in Windows is to put data in %appdata% which is what I do.

> Putting the document folders inside the installation folder would be the more logical solution : no one in schools use openboard for different users on the same computer.

Except every school at Geneva...

It doesn't mean I'll never look at your issue but please understand that we have very limited resources and I must prioritize on what what concerns the users I work for.

**Comment by SeeDborg** [comment]
I understand very much, I hope I did not seem aggressive in my comments and I apologize if that is the case.

It would be a very useful feature for the teachers I work with in France.
Being able to use a relative path for documents (wihout the recursive bug ^^) would be a nice fix for now.

Thanks again for your work.

**Comment by letsfindaway** [comment]
> I understand very much, I hope I did not seem aggressive in my comments and I apologize if that is the case.

Don't worry. It is ok with me. 
 
> It would be a very useful feature for the teachers I work with in France. Being able to use a relative path for documents (wihout the recursive bug ^^) would be a nice fix for now.

I think a fix for the "recursive recording" would be desirable. But at the same time I would like to define the base directory for relative paths in `DataDirectory`. And for sure it will not be the OpenBoard installation directory for the reasons stated above. So this will not help you.

I don't have Windows here and my knowledge about this is limited. Just to make my proposal a little bit easier: It is sufficient to **append** an additional `[App]` section with the new data directory instead of replacing a potentially existing entry:

```
[App]
DataDirectory=E:/path/on/usb/stick
```

Note the use of forward slashes!!! Alternatively double backslashes.

I think an experienced Windows user could help you writing such a script:

```
set scriptpath=%~dp0
# somehow replace backslashes by forward slashes
# append the desired document path
# append [App] to the OpenBoardUser.config file
# append DataDirectory=<path> to the OpenBoardUser.config file
# start OpenBoard
# when OpenBoard terminates,
# remove the DataDirectory line from the OpenBoardUser.config file
```

Place this script on your USB drive and start OpenBoard using that.

This solution would not need any modifications of OpenBoard and (hopefully) serves your specific use case.

You just won't get such a script from me, as I am a Linux only user.

**Comment by letsfindaway** [comment]
**EDIT** to my previous post: `DataDirectory` is not processed from the `OpenBoardUser.config` file. It must be changed in the `OpenBoard.config` file provided by the installation.

**Comment by SeeDborg** [comment]
I just installed another software and I think I may have found a solution for the problem we talked about.

Some softwares offer a choice when installed : 

- For all users (installs in C:/Programs
- Just for me (installs in the user folder)

Wouldn't it be the perfect solution ?

**Comment by letsfindaway** [comment]
> I just installed another software and I think I may have found a solution for the problem we talked about.
> 
> Some softwares offer a choice when installed :
> 
> * For all users (installs in C:/Programs
> * Just for me (installs in the user folder)
> 
> Wouldn't it be the perfect solution ?

I don't think that this would change anything for your problem. No matter where the program is installed, on Windows the document folder is per default in `C:/Users/<USER>/AppData/Local/<APPNAME>` or, if `App/DataDirectory` is specified, in that absolute path.
